okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/ConcurrencyLimiters.java

```diff
@@ -0,0 +1,124 @@
+import com.palantir.tracing.okhttp3.OkhttpTraceInterceptor;
+import com.palantir.tritium.metrics.registry.MetricName;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import okhttp3.Request;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("DesignForExtension")
+class ConcurrencyLimiters {
+    private static final Logger log = LoggerFactory.getLogger(ConcurrencyLimiters.class);
+    private static final Duration DEFAULT_TIMEOUT = Duration.ofMinutes(1);
+    private static final Void NO_CONTEXT = null;
+    private static final String FALLBACK = "";
+    private static final MetricName SLOW_ACQUIRE =
+            MetricName.builder().safeName("conjure-java-client.qos.request-permit.slow-acquire").build();
+
+    private final Timer slowAcquire;
+    private final ConcurrentMap<String, Limiter<Void>> limiters = new ConcurrentHashMap<>();
+    private final Duration timeout;
+
+    @VisibleForTesting
+    ConcurrencyLimiters(TaggedMetricRegistry taggedMetricRegistry, Duration timeout) {
+        this.slowAcquire = taggedMetricRegistry.timer(SLOW_ACQUIRE);
+        this.timeout = timeout;
+    }
+
+    ConcurrencyLimiters(TaggedMetricRegistry taggedMetricRegistry) {
+        this(taggedMetricRegistry, DEFAULT_TIMEOUT);
+    }
+
+    /**
+     * Blocks until the request should be allowed to proceed.
+     * Caller must notify the listener to release the permit.
+     */
+    Limiter.Listener acquireLimiter(Request request) {
+        return acquireLimiter(limiterKey(request));
+    }
+
+    @VisibleForTesting
+    Limiter.Listener acquireLimiter(String name) {
+        Limiter<Void> limiter = limiters.computeIfAbsent(name, key -> newLimiter());
+        Optional<Limiter.Listener> listener = limiter.acquire(NO_CONTEXT);
+        return listener.orElseGet(() -> {
+            if (Thread.currentThread().isInterrupted()) {
+                throw new RuntimeException("Thread was interrupted");
+            }
+            log.warn("Timed out waiting to get permits for concurrency. In most cases this would indicate "
+                            + "some kind of deadlock. We expect that either this is caused by not closing response "
+                            + "bodies (there should be OkHttp log lines indicating this), or service overloading.",
+                    SafeArg.of("timeout", timeout));
+            limiters.replace(name, limiter, newLimiter());
+            return acquireLimiter(name);
+        });
+    }
+
+    private Limiter<Void> newLimiter() {
+        Limiter<Void> limiter = new InstrumentedLimiter(SimpleLimiter.newBuilder()
+                .limit(new RemotingWindowedLimit(VegasLimit.newDefault()))
+                .build(), slowAcquire);
+        return RemotingBlockingLimiter.wrap(limiter, timeout);
```

Comment:
can we call this ConjureBlockingLimiter?

Reply:
Done!